add ability to get issues from multiple repos

mancdaz opened the pull request:
- add flag to pass in multiple repo names
- return all labels for all repos and use for headers
- write headers once
- process each repo in turn and write out issues

mancdaz committed:
add ability to get issues from multiple repos

- add flag to pass in multiple repo names
- return all labels for all repos and use for headers
- write headers once
- process each repo in turn and write out issues

git-harry commented on gh2/csv.py:
```diff
@@ -59,8 +59,11 @@ def make_parser():
+        help='Repository to retrieve issues from (e.g., rcbops/rpc-openstack). '
+             'Multiple invocations of this flag mean issues from all repos will '
+             'be collated into a single output file.',
+        action='append', dest='repositories', default=[],
```
I think you want required=True here.

mancdaz replied:
fixed

git-harry commented on gh2/csv.py:
```diff
@@ -245,14 +251,39 @@ def main():
         'Milestone'
```
It's not possible to filter by repo

mancdaz replied:
I don't understand this comment here

git-harry replied:
There is no column for repo. This means if you generate a file with data from multiple repos there is no easy way to filter out one or more of the repos. I don't know if that is likely to cause an issue.

mancdaz replied:
this isn't a major concern right now, but can be added as a feature if needed later.

git-harry commented on gh2/csv.py:
```diff
@@ -174,12 +177,17 @@ def normalize_sequential_dates(issue_list):
+    with open(filename, 'w+') as fd:
+        writer = csv.writer(fd)
+        writer.writerow(headers)
+        fd.close()
```
You don't need this line, it is done automatically when leaving the context.

mancdaz replied:
yup, for some reason it was not behaving properly when I first tested it, but you're right.

git-harry commented on gh2/csv.py:
```diff
@@ -59,8 +59,11 @@ def make_parser():
-        'repository',
-        help='Repository to retrieve issues from (e.g., rcbops/rpc-openstack)',
+        '--repo',
+        help='Repository to retrieve issues from (e.g., rcbops/rpc-openstack). '
```
You might want to run pep8 on the file, there are a few things to clean up such as these lines which are too long.

mancdaz replied:
fixed

git-harry commented on gh2/csv.py:
```diff
@@ -206,25 +214,23 @@ def set_headers(repo, labels=None):
+                labels.append(repo_label)
+    return sorted((labels), key = lambda l: l.name)
+
+def get_repo_labels(repo):
```
This function isn't really doing anything, I'd get rid of it.

mancdaz replied:
done

git-harry commented on gh2/csv.py:
```diff
@@ -206,25 +214,23 @@ def set_headers(repo, labels=None):
+        repo = get_repo(repo_owner, repo_name, token)
+        repo_labels = get_repo_labels(repo)
+        for repo_label in repo_labels:
+           if repo_label.name not in [l.name for l in labels]:
```
I don't think there is any need to keep the label objects, if you just use the names you can use a set which would allow you to replace this for-loop with
```python
all_labels.update(repo_labels)
```

mancdaz replied:
the list of label objects that this function returns (as additional_labels) can be/is reused elsewhere so I didn't want to discard the objects inside the function

git-harry commented on gh2/csv.py:
```diff
@@ -206,25 +214,23 @@ def set_headers(repo, labels=None):
+        for repo_label in repo_labels:
+           if repo_label.name not in [l.name for l in labels]:
+                labels.append(repo_label)
+    return sorted((labels), key = lambda l: l.name)
```
There is no need for brackets around labels and the `=` shouldn't have whitespace around it.

mancdaz replied:
fixed

git-harry commented on gh2/csv.py:
```diff
@@ -245,14 +251,39 @@ def main():
+    args = parser.parse_args()
+
+    filename = args.output_file
+    repositories = args.repositories
```
Instead of passing round the repositories, why not make a list of the repo objects here. That save having to pass the token into get_all_labels and saves having to run get_repo in two places.

mancdaz replied:
done

git-harry commented on gh2/csv.py:
```diff
@@ -206,25 +214,23 @@ def set_headers(repo, labels=None):
+    labels = []
+    for repository in repositories:
+        repo_owner, repo_name = repository.split('/', 1)
+        repo = get_repo(repo_owner, repo_name, token)
```
As mentioned elsewhere, I think it would be better if repositories was a list of the repo objects. This would save having to pass the token into this function and call get_repo here.

mancdaz replied:
done

git-harry reviewed (changes requested):
Seems to work okay, a few comments in the code.

mancdaz committed:
additional fixes on multi-repo feature

- flake8 fixes
- removed some unused args
- remove unused functions
- build list of repo objects in main() for reuse
-

mancdaz commented:
Thanks for the review @git-harry - changes made and ready for review again

git-harry commented on gh2/csv.py:
```diff
@@ -207,24 +209,26 @@ def set_headers(repo, labels=None):
 
-def main():
-    parser = make_parser()
+def get_all_labels(repositories):
```
I don't think it make sense to keep the label objects because they include information specific to the repo from which they came. So, some label objects may relate to one repo and other objects to others even though the label names may relate to multiple repos.

I would still be inclined towards using something like the following:
```python
def get_all_labels(repositories):                                                 
    labels = set()                                                                 
    for repo in repositories:                                                      
        labels.update(l.name for l in repo.labels())                               
    return sorted(labels)
```

Note, this change would require line 208 being updated to:
```python
        headers.extend('Label: ' + label for label in labels)
```

mancdaz replied:
@git-harry it would also require change to the ``issue_to_dict`` function which also uses ``additional_labels`` as a list of label objects, but I take the point about the label objects from different repos not being the same even if they are the same name, so I'll try and work it out

git-harry replied:
Ah, I'd missed the issue_to_dict bit but now I look at it, this patch introduces a bug. If you look at line 90 it tests for the label using the label object but that wouldn't work if a label could be from either repo.

mancdaz replied:
@git-harry yeah, it's currently broken. I'll need to fix this later, either by modifying ```issue_to_dict``` to work with a passed in list of label names instead of objects, or working to extract the names within the function for comparison.

git-harry commented on gh2/csv.py:
```diff
@@ -245,14 +249,34 @@ def main():
+        headers=headers
     )
+
+    # repo loop
```
This comment doesn't add any additional information.

mancdaz replied:
removed

git-harry commented on gh2/csv.py:
```diff
@@ -207,24 +209,26 @@ def set_headers(repo, labels=None):
+def get_token():
     token = os.environ.get('GITHUB_TOKEN')
     if token is None:
         parser.exit(status=1,
```
parser is not passed into the function.

mancdaz replied:
fixed

mancdaz committed:
Only pass around a list of label names instead of objects

Since we only use the label names in all places, let's just pass around
an iterable (in this case a set) of label names for the additional
labels that we might want to include in the output

mancdaz commented:
ok @git-harry I've made it so that we only get, and pass around, a set of label names. The appropriate adjustments have been made in the places where a list of label objects were expected.

git-harry commented:
Looks to work okay. I gave the output a quick test in the graphing tool and it didn't seem to be affected by the duplicate IDs so I think this is ready.

mancdaz commented:
Thanks @git-harry
